src/detect-tag-pcap.c

```diff
@@ -0,0 +1,340 @@
+    }
+}
+
+/**
+ * \brief Dumps content of a TcpSegment to specified pcap output file.
+ * \param tcpSegment to be dumped to pcap.
+ * \param dump_handle pcap_dumper location.
+ * \bool client direction of segment (to client or to server).
+ */
+static void DumpTcpSegment(TcpSession *session, TcpSegment *seg,
+        pcap_dumper_t *dump_handle, bool client)
+{
+    struct pcap_pkthdr pcap_hdr;
+    uint32_t packet_len = seg->pktlen;
+    uint32_t payload_len = seg->payload_len;
+    uint32_t packet_header_len = packet_len - payload_len;
+
+    pcap_hdr.ts.tv_sec = seg->ts.tv_sec;
+    pcap_hdr.ts.tv_usec = seg->ts.tv_usec;
+    pcap_hdr.caplen = packet_len;
+    pcap_hdr.len = packet_len;
+
+    uint8_t *packet_data = (uint8_t*) SCMalloc(packet_len);
+    if (packet_data != NULL && seg->pkt_hdr != NULL) {
+        memcpy(packet_data, seg->pkt_hdr, packet_header_len);
+        if (client) {
+            memcpy(packet_data + packet_header_len, session->client.sb.buf +
+                seg->sbseg.stream_offset, payload_len);
+        } else {
+            memcpy(packet_data + packet_header_len, session->server.sb.buf +
+                seg->sbseg.stream_offset, payload_len);
+        }
+        pcap_dump((u_char *) dump_handle, &pcap_hdr, packet_data);
+    } else {
+        return;
+    }
+    SCFree(packet_data);
+}
+
+/**
+ * \brief A helper function that finds and logs the oldest TCP segment in the
+ *  session to begin processing the whole session.
+ * \param session to be dumped to pcap.
+ * \param dump_handle pcap_dumper location.
+ * \retval The oldest tcp segment if it exists, otherwise NULL.
+ */
+static TcpSegment *LogOldestTcpSegment(TcpSession *session, pcap_dumper_t
+        *dump_handle)
+{
+    TcpSegment *server_node = session->server.seg_tree.rbh_root;
+    TcpSegment *client_node = session->client.seg_tree.rbh_root;
+
+    if (client_node == NULL || server_node == NULL) {
+        return NULL;
+    }
+
+    while (RB_LEFT(server_node, rb) != NULL) {
+        server_node = RB_LEFT(server_node, rb);
+    }
+    while (RB_LEFT(client_node,rb) != NULL) {
+        client_node = RB_LEFT(client_node,rb);
+    }
+    if (server_node->pcap_cnt < client_node->pcap_cnt) {
+        DumpTcpSegment(session, server_node, dump_handle, false);
+        return server_node;
+    } else {
+        DumpTcpSegment(session, client_node, dump_handle, true);
+        return client_node;
+    }
+}
+
+/**
+ * \brief Searches through the RB trees of tcp segments in both directions
+ *  (to server and to client) for the tcp segment with the specified pcap_cnt
+ *  value. If found, the segment is logged to output pcap file.
+ * \param session to be dumped to pcap.
+ * \param dump_handle pcap_dumper location.
+ * \param pcap_cnt A value added to tcp segment to keep track of the original
+ *  packet processing order.
+ * \return The tcp segment if it exists. Otherwise NULL.
+ */
+static TcpSegment *FindAndLogTcpSegment(TcpSession *session, pcap_dumper_t
```

Comment:
I don't understand this logic of finding the oldest and then searching. Can we just walk using TCPSEG_RB_NEXT? That avoids a search for each segment. If the order needs to be preserved you should be able to walk both in parallel and use pcap_cnt to dump and progress the correct direction first?